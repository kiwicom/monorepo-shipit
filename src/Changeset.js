// @flow

type Diff = {|
  +path: string,
  +body: string,
|};

opaque type ChangesetData = {|
  +id: string,
  +timestamp: number,
  +author: string,
  +subject: string,
  +message: string,
  +diffs: Set<Diff>,
|};

export default class Changeset {
  id: string;
  message: string;
  diffs: Set<Diff>;

  // TODO: change the order (?)
  constructor(id: string, diffs: Set<Diff>, message: string) {
    this.id = id;
    this.message = message;
    this.diffs = diffs;
  }

  getID = () => {
    return this.id;
  };

  getMessage = () => {
    return this.message;
  };

  withMessage = (message: string): Changeset => {
    return new Changeset(this.id, this.diffs, message);
  };
}
