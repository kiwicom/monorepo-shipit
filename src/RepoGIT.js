// @flow

import fs from 'fs';
import path from 'path';
import { invariant } from '@kiwicom/js';
import { ChildProcess } from '@kiwicom/monorepo';

import parsePatch from './parsePatch';
import parsePatchHeader from './parsePatchHeader';
import splitHead from './splitHead';
import Changeset from './Changeset';

/**
 * This is our monorepo part - source of exports.
 */
export interface SourceRepo {
  findDescendantsPath(
    baseRevision: string,
    headRevision: string,
    roots: Set<string>,
  ): $ReadOnlyArray<string>;
  getChangesetFromID(revision: string): Changeset;
  getNativePatchFromID(revision: string): string;
  getNativeHeaderFromIDWithPatch(revision: string, patch: string): string;
  getChangesetFromExportedPatch(header: string, patch: string): Changeset;
}

/**
 * Exported repository containing `kiwicom-source-id` handles.
 */
export interface DestinationRepo {
  findLastSourceCommit(roots: Set<string>): string;
  renderPatch(changeset: Changeset): string;
  commitPatch(changeset: Changeset): void;
  push(): void;
}

export default class RepoGIT implements SourceRepo, DestinationRepo {
  localPath: string;

  constructor(localPath: string) {
    this.localPath = localPath;
  }

  _gitCommand = (args: $ReadOnlyArray<string>, options: Object) => {
    invariant(
      fs.existsSync(path.join(this.localPath, '.git')),
      '%s is not a GIT repo.',
      this.localPath,
    );
    return ChildProcess.executeSystemCommand(
      'git',
      [
        '--no-pager',
        ...args.filter(arg => arg !== ''), // TODO: removes empty strings - we should have some escaping but there is nothing in Node.js (?)
      ],
      { cwd: this.localPath, ...options },
    );
  };

  push() {
    this._gitCommand(['push', 'origin', 'master']);
  }

  findLastSourceCommit = (roots: Set<string>): string => {
    const rawLog = this._gitCommand([
      'log',
      '-1',
      '--grep',
      '^kiwicom-source-id: [a-z0-9]\\+$',
      ...roots,
    ]);
    const match = rawLog
      .trim()
      .match(/kiwicom-source-id: (?<commit>[a-z0-9]+)$/m);
    return match?.groups?.commit ?? 'd30a77bd2fe0fdfe5739d68fc9592036e94364dd'; // very first commit in incubator/universe
  };

  getNativePatchFromID = (revision: string): string => {
    return this._gitCommand([
      'format-patch',
      '--no-renames',
      '--no-stat',
      '--stdout',
      '--full-index',
      '--format=', // contain nothing but the code changes
      '-1',
      revision,
    ]);
  };

  getNativeHeaderFromIDWithPatch = (
    revision: string,
    patch: string,
  ): string => {
    const fullPatch = this._gitCommand([
      'format-patch',
      '--no-renames',
      '--no-stat',
      '--stdout',
      '--full-index',
      '-1',
      revision,
    ]);
    if (patch.length === 0) {
      // this is an empty commit, so everything is the header
      return fullPatch;
    }
    return fullPatch.replace(patch, '');
  };

  getChangesetFromID = (revision: string): Changeset => {
    const patch = this.getNativePatchFromID(revision);
    const header = this.getNativeHeaderFromIDWithPatch(revision, patch);
    const changeset = this.getChangesetFromExportedPatch(header, patch);
    console.warn(`Filtering changeset for: ${revision}`); // eslint-disable-line no-console
    return changeset.withID(revision);
  };

  getChangesetFromExportedPatch = (
    header: string,
    patch: string,
  ): Changeset => {
    const changeset = parsePatchHeader(header);
    const diffs = new Set();
    for (const hunk of parsePatch(patch)) {
      const diff = this.parseDiffHunk(hunk);
      if (diff !== null) {
        diffs.add(diff);
      }
    }
    return changeset.withDiffs(diffs);
  };

  parseDiffHunk = (hunk: string) => {
    const [head, tail] = splitHead(hunk, '\n');
    const match = head.match(/^diff --git [ab]\/(?:.*?) [ab]\/(?<path>.*?)$/);
    if (!match) {
      return null;
    }
    const path = match.groups?.path;
    invariant(path != null, 'Cannot parse path from the hunk.');
    return { path, body: tail };
  };

  // TODO: originally `findNextCommit` - pls reconsider
  findDescendantsPath = (
    baseRevision: string,
    headRevision: string,
    roots: Set<string>,
  ): $ReadOnlyArray<string> => {
    const log = this._gitCommand([
      'log',
      '--reverse',
      '--ancestry-path',
      '--no-merges',
      '--pretty=tformat:%H',
      baseRevision + '..' + headRevision,
      ...roots,
    ]);

    // return descendant hashes
    return log.trim().split('\n');
  };

  commitPatch = (changeset: Changeset): void => {
    const diff = this.renderPatch(changeset);
    // TODO: catch an error here and run `am --abort`
    this._gitCommand(['am', '--keep-non-patch', '--keep-cr'], {
      // stdin, stdout, stderr
      stdio: ['pipe', 'inherit', 'inherit'],
      input: diff,
    });
  };

  renderPatch = (changeset: Changeset): string => {
    let renderedDiffs = '';
    for (const diff of changeset.getDiffs()) {
      const path = diff.path;
      renderedDiffs += `diff --git a/${path} b/${path}
${diff.body}`;
    }

    // Mon Sep 17 is a magic date used by format-patch to distinguish from real mailboxes
    // see: https://git-scm.com/docs/git-format-patch
    return `From ${changeset.getID()} Mon Sep 17 00:00:00 2001
From: ${changeset.getAuthor()}
Date: ${changeset.getTimestamp()}
Subject: [PATCH] ${changeset.getSubject()}

${changeset.getDescription()}

${renderedDiffs}
--
2.21.0
`;
  };
}
