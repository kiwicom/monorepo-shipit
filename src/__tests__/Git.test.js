// @flow

import RepoGIT from '../RepoGIT';
import Changeset from '../Changeset';

jest.mock('fs', () => ({
  // workarounding fake Git repo paths
  existsSync: () => true,
}));

jest.mock('@kiwicom/monorepo-utils', () => {
  return {
    ShellCommand: class {
      stdout = null;
      constructor(path, git, noPager, ...commandArray) {
        const command = commandArray.toString();
        switch (command) {
          case 'rev-list,--max-parents=0,HEAD':
            this.stdout =
              path === 'mocked_repo_path_1'
                ? 'd30a77bd2fe0fdfe5739d68fc9592036e94364dd'
                : '0ca71b3737cbb26fbf037aa15b3f58735785e6e3\n' +
                  '16d6b8ab6fd7f68bfd9f4d312965cb99e8ad911b\n' +
                  'cb07fc2a29c86d1bc11f5415368f778d25d3d20a\n' +
                  '161332a521fe10c41979bcd493d95e2ac562b7ff\n' +
                  '1db95b00a2d2a001fd91cd860a71c639ea04eb53\n' +
                  '2744b2344dc42fa2a1ddf17f4818975cd48f6d42\n' +
                  'e83c5163316f89bfbde7d9ab23ca2e25604af290'; // we should be interested only in this one
            break;
          default:
            throw new Error(`There is no available dataset for command: ${command}`);
        }
      }

      getStdout() {
        return this.stdout;
      }

      runSynchronously() {
        return this;
      }

      setEnvironmentVariables() {
        return this;
      }
    },
  };
});

it('renders patch as expected', () => {
  const repo = new RepoGIT('mocked_repo_path_1');
  const changeset = new Changeset()
    .withID('mocked_id')
    .withTimestamp('mocked_timestamp')
    .withAuthor('mocked_author')
    .withSubject('mocked_subject')
    .withDescription('mocked_description')
    .withDiffs(
      new Set([
        { path: 'mocked_path1', body: 'mocked_body1\n' },
        { path: 'mocked_path2', body: 'mocked_body2\n' },
      ]),
    );
  expect(repo.renderPatch(changeset)).toMatchSnapshot();
});

describe('findFirstAvailableCommit', () => {
  beforeEach(() => {
    jest.spyOn(console, 'warn').mockImplementation(() => {});
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('returns first available revision when single root available', () => {
    const repo = new RepoGIT('mocked_repo_path_1');
    expect(repo.findFirstAvailableCommit()).toBe('d30a77bd2fe0fdfe5739d68fc9592036e94364dd');
  });

  it('returns first available revision when multiple roots available', () => {
    const repo = new RepoGIT('mocked_repo_path_2');
    expect(repo.findFirstAvailableCommit()).toBe('e83c5163316f89bfbde7d9ab23ca2e25604af290');
  });
});
