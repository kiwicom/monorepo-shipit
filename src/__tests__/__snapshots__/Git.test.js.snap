// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders patch as expected 1`] = `
"From mocked_id Mon Sep 17 00:00:00 2001
From: mocked_author
Date: mocked_timestamp
Subject: [PATCH] mocked_subject

mocked_description

diff --git a/mocked_path1 b/mocked_path1
mocked_body1
diff --git a/mocked_path2 b/mocked_path2
mocked_body2

--
2.21.0
"
`;
